### PR TITLE
Prototype PDBError and add error.rs

### PR DESCRIPTION
**Pull request by tinoism**
This draft PR provides a MVP prototype for PDBError and adds error.rs to change the return type from `Option` to `Result`. Specifically, I changed the return type of `get_compact_symbol_table` in pdb.rs from `Option<CompactSymbolTable>` to `CompactSymbolResult<CompactSymbolTable>`. 

I added code in error.rs so as to show examples of translating `pdb_crate::Error` to `CompactSymbolError`. 

@mstange Let me know if this is the right direction to go. The error codes have not been specified yet (with UnknownError as the placeholder for now).

References [Bug 1563305](https://bugzilla.mozilla.org/show_bug.cgi?id=1563305).

### CODE REVIEW

**Comment by tinoism** [comment]
Oops, I meant to flag this PR as draft. It's currently not ready to be merged yet.

File: src/error.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
+use goblin::error::Error as GoblinError;
+use pdb_crate::Error as PDBError;
+pub type CompactSymbolResult<T> = Result<T, CompactSymbolError>;
```
Comment (mstange) [inline src/error.rs]: You can call this `Result`:

```rust
type Result<T> = std::result::Result<T, GetSymbolsError>;
```

File: src/lib.rs [hunk 2]
```diff
@@ -50,32 +53,30 @@ fn get_compact_symbol_table_impl(
...
+            return elf::get_compact_symbol_table(binary_data, breakpad_id);
+        }
+        Hint::Mach(_) => {
+            return macho::get_compact_symbol_table(binary_data, breakpad_id);
+        }
+        Hint::MachFat(_) => {
+            let multi_arch = mach::MultiArch::new(binary_data)?;
+            let fat_arch = multi_arch.iter_arches().filter(|x| x.is_ok()).next().ok_or(CompactSymbolError::NoneValueError)?;
+            let arch_slice = fat_arch?.slice(binary_data);
+            return macho::get_compact_symbol_table(arch_slice, breakpad_id);
```
Comment (mstange) [inline src/lib.rs]: You changed the meaning of this code: Before, `get_compact_symbol_table` was called for every arch until it succeeded, and now it's just called for the first arch. We need the original meaning, for the following purpose:

Some macOS system libraries contain two architectures in their MultiArch file: x86_64 and x86_64h. (The h is for the "Haswell" CPU architecture.) The two architectures have different breakpadIDs. The only way this code knows which of the two to get symbols for is based on the `breakpad_id` parameter. For example, the first arch in the MultiArch container might parse successfully but then error out because of the breakpad ID mismatch. Then we try to parse the second arch, find out that its breakpad ID is the one we were looking for, and successfully use that second arch.

File: src/error.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+#[derive(Fail, Debug)]
+pub enum CompactSymbolError {
+    #[fail(display = "Strings do not match")]
+    UnmatchedString,
```
Comment (mstange) [inline src/error.rs]: Let's be even more specific here and make this error describe the exact thing that mismatched, which is the breakpad ID. (I think we shouldn't try to find "categories" of errors - just give a separate enum value to every possible error occurrence, for maximum clarity.)
E.g. display = "Breakpad ID does not match", UnmatchedBreakpadID

File: src/error.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    IOError,
+
+    #[fail(display = "Invalid string: {}", _0)]
+    InvalidStringError(&'static str),
+
+    #[fail(display = "Invalid result: {}", _0)]
+    InvalidResultError(&'static str),
+
+    #[fail(display = "None value error")]
+    NoneValueError,
```
Comment (mstange) [inline src/error.rs]: Invalid string, invalid result, and none value error are a bit too generic for my taste. Let's make a different name for each specific thing that can go wrong.

File: src/error.rs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    #[fail(display = "goblin error: {}", _0)]
+    GoblinError(GoblinError),
+}
+
+impl From<PDBError> for CompactSymbolError {
+    fn from(err: PDBError) -> CompactSymbolError {
+        match err {
+            PDBError::UnimplementedFeature(_) => CompactSymbolError::UnimplementedFeature,
```
Comment (mstange) [inline src/error.rs]: For this one, I think it would be good to indicate that something was unimplemented in PDB specifically. So maybe it's easier to just take the default case for `PDBError::UnimplementedFeature` and simply wrap it in a `CompactSymbolError::PDBError`.

**Review by mstange — COMMENTED** [review]
Looking pretty good so far!

I left some comments about the error variant names. But I have a more general comment: There are many places where you're adding an `ok_or` call. Could you avoid some of those additions by adding more `impl From<SomeError> for CompactSymbolError` sections? If not, why not?

File: src/pdb.rs [hunk 6]
```diff
@@ -7,26 +8,22 @@ use std::io::Cursor;
...
-        format!("{:X}", info.guid.to_simple()),
-        info.age
-    );
+    let mut pdb = PDB::open(pdb_reader)?;
+    let info = pdb.pdb_information()?;
+    let pdb_id = format!("{}{:x}", format!("{:X}", info.guid.to_simple()), info.age);
 
     if pdb_id != breakpad_id {
-        return None;
+        return Err(CompactSymbolError::UnmatchedString);
```
Comment (mstange) [inline src/pdb.rs]: When you have renamed this error variant to `UnmatchedBreakpadID`, I think it would be a good idea to store both breakpad IDs in it, so that the error can be displayed as "expected ID ..., found ID ...".

File: src/error.rs (hunk 1 above)
Comment (tinoism) [reply src/error.rs]: With that type declaration, we wouldn't be able to abstract `Result` to other future classes of errors.

File: src/error.rs (hunk 4 above)
Comment (tinoism) [reply src/error.rs]: Ah yes. NoneValueError is actually propagating `std::convert::From<std::option::NoneError>`, but I cannot directly propagate that because of `use of unstable library feature 'try_trait`. I tried to pass on different headers like `unstable` but still couldn't get it to work. So I decide to just have `NoneValueError` there.

File: src/error.rs [hunk 7]
```diff
@@ -4,23 +4,17 @@ pub type CompactSymbolResult<T> = Result<T, CompactSymbolError>;
 
 #[derive(Fail, Debug)]
 pub enum CompactSymbolError {
-    #[fail(display = "Strings do not match")]
-    UnmatchedString,
+    #[fail(display = "Unmatched strings: Expected {}, but received {}", _0, _1)]
+    UnmatchedString(String, String),
```
Comment (tinoism) [inline src/error.rs]: Here I added two parameters to the `UnmatchedString` error. Where the first param is the expected value and second is actual. Hope this pins down what parameter values are causing the error to occur.

File: src/error.rs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    UnimplementedFeature,
+
+    #[fail(display = "IO Error")]
+    IOError,
+
+    #[fail(display = "Unknown error")]
+    UnknownError,
+
+    #[fail(display = "Invalid string: {}", _0)]
+    InvalidStringError(&'static str),
```
Comment (tinoism) [inline src/error.rs]: For `InvalidStringError` and `InvalidResultError`, I collapse it to one single `InvalidInputError` which takes in a  error message parameter describing what parameter cannot be parsed specifically so to not make it generic.

File: src/lib.rs [hunk 9]
```diff
@@ -92,32 +95,32 @@ fn get_compact_symbol_table_impl(
...
-                    if let Some(table) = macho::get_compact_symbol_table(arch_slice, breakpad_id) {
-                        return Some(table);
-                    }
+    match goblin::peek(&mut reader)? {
+        Hint::Elf(_) => elf::get_compact_symbol_table(binary_data, breakpad_id),
+        Hint::Mach(_) => macho::get_compact_symbol_table(binary_data, breakpad_id),
+        Hint::MachFat(_) => {
+            let multi_arch = mach::MultiArch::new(binary_data).ok();
+            for fat_arch in multi_arch
+                .unwrap()
```
Comment (mstange) [inline src/lib.rs]: You added an `.unwrap()` here - can you propagate the error instead?

File: src/lib.rs [hunk 10]
```diff
@@ -92,32 +95,32 @@ fn get_compact_symbol_table_impl(
...
+                    return Ok(table);
                 }
             }
-            Hint::PE => {
-                return pdb::get_compact_symbol_table(debug_data, breakpad_id);
-            }
-            _ => {}
+            Err(GetSymbolsError::InvalidInputError(
+                "Incompatible system architecture",
+            ))
```
Comment (mstange) [inline src/lib.rs]: I think we can do a bit better for this error. Because we're iterating over multiple sub-binaries here, and discard the errors from failing calls to `macho::get_compact_symbol_table`, `InvalidInputError` might not be entirely accurate.
How about this: For the first call to `macho::get_compact_symbol_table`, if that call returns an error, store that error in a `first_error: Option<GetSymbolsError>` variable. Then, if we get to the end of the loop (meaning all sub-binaries failed in some way), and `first_error` is not `None`, return that error. Otherwise return `InvalidInputError` as you're doing now.

File: src/macho.rs [hunk 11]
```diff
@@ -1,10 +1,16 @@
...
-
-pub fn get_compact_symbol_table(buffer: &[u8], breakpad_id: &str) -> Option<CompactSymbolTable> {
-    let macho_file = MachOFile::parse(buffer).ok()?;
-    if format!("{:X}0", macho_file.mach_uuid()?.simple()) != breakpad_id {
-        return None;
+pub fn get_compact_symbol_table(buffer: &[u8], breakpad_id: &str) -> Result<CompactSymbolTable> {
+    let macho_file = MachOFile::parse(buffer).or(Err(GetSymbolsError::InvalidInputError(
+        "Could not parse Mach-O header",
+    )))?;
+    let macho_id = format!("{:X}0", macho_file.mach_uuid().unwrap().simple());
```
Comment (mstange) [inline src/macho.rs]: There's another unwrap call here.

File: src/error.rs [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+pub type Result<T> = std::result::Result<T, GetSymbolsError>;
+
+#[derive(Debug)]
+pub enum GetSymbolsError {
+    UnmatchedBreakpadId(String, String),
+    PDBError(String),
+    IOError,
+    InvalidInputError(&'static str),
+    GoblinError(String),
```
Comment (mstange) [inline src/error.rs]: In a previous patch, you were wrapping the original errors from the various crates and not converting them to strings. The change to strings was necessary because you were trying to Serialize this error type directly. But now you have a separate error type for serialization (which I think is good), so you don't have to stringify early any more. Could you go back to wrapping the original error types? I think that would be a bit cleaner. Thanks!

File: src/pdb.rs [hunk 13]
```diff
@@ -30,29 +34,28 @@ pub fn get_compact_symbol_table(pdb_data: &[u8], breakpad_id: &str) -> Option<Co
...
+        while let Some(module) = modules.next()? {
+            let info = pdb.module_info(&module)?;
+            let mut symbols = info.symbols()?;
+            while let Some(symbol) = symbols.next()? {
                 if let Ok(SymbolData::Procedure(ProcedureSymbol { offset, .. })) = symbol.parse() {
-                    let name = symbol.name().ok()?;
+                    let name = symbol.name()?;
                     hashmap
-                        .entry(offset.to_rva(&addr_map)?.0)
+                        .entry(offset.to_rva(&addr_map).unwrap().0)
```
Comment (mstange) [inline src/pdb.rs]: Can you avoid this unwrap(), too?

File: src/elf.rs [hunk 14]
```diff
@@ -1,18 +1,25 @@
...
 const UUID_SIZE: usize = 16;
 const PAGE_SIZE: usize = 4096;
 
-pub fn get_compact_symbol_table(buffer: &[u8], breakpad_id: &str) -> Option<CompactSymbolTable> {
-    let elf_file = ElfFile::parse(buffer).ok()?;
-    let elf_id = get_elf_id(&elf_file, buffer)?;
-    if format!("{:X}0", elf_id.simple()) != breakpad_id {
-        return None;
+pub fn get_compact_symbol_table(buffer: &[u8], breakpad_id: &str) -> Result<CompactSymbolTable> {
+    let elf_file = ElfFile::parse(buffer).or(Err(GetSymbolsError::InvalidInputError(
```
Comment (mstange) [inline src/elf.rs]: Could you use `.map_err` in all the places where you're using `.or`? That will avoid creating the error object in the passing case.

File: src/elf.rs [hunk 15]
```diff
@@ -1,18 +1,25 @@
...
+        "Could not parse ELF header",
+    )))?;
+    let elf_id = get_elf_id(&elf_file, buffer)
+        .ok_or(GetSymbolsError::InvalidInputError("id cannot be read"))?;
```
Comment (mstange) [inline src/elf.rs]: And use `.ok_or_else` in the places where you're using `.ok`, for the same reason.

**Review by mstange — CHANGES_REQUESTED** [review]
This is looking really good now! I still have some requests for changes, though.

File: src/lib.rs (hunk 9 above)
Comment (tinoism) [reply src/lib.rs]: We'd bump into the error we discussed earlier-- `NoneError is not implemented ...`. I think we'd need to have option here...

File: src/macho.rs (hunk 11 above)
Comment (tinoism) [reply src/macho.rs]: Similarly, we'd have the error ``` the trait `std::convert::From<std::option::NoneError>` is not implemented for `error::GetSymbolsError` ``` if we use `?` instead of `.unwrap()`

File: src/pdb.rs (hunk 13 above)
Comment (tinoism) [reply src/pdb.rs]: We'd then have``` the trait `std::convert::From<std::option::NoneError>` is not implemented for `error::GetSymbolsError` ```

File: src/lib.rs (hunk 9 above)
Comment (mstange) [reply src/lib.rs]: Oh, that's because you're calling it on an `Option`, which you obtained by calling `.ok()` in the line above. Right?
What happens if you remove both the unwrap and the ok, and put the question mark where the `.ok()` was?

File: src/error.rs (hunk 12 above)
Comment (tinoism) [reply src/error.rs]: If we don't stringify early, we would be converting `PDBError` to a generic `GetSymbolsError:: PDBError` with loss of the `PDBError` message. In that case, we can't pinpoint what PDB message has been fired?

File: src/macho.rs (hunk 11 above)
Comment (mstange) [reply src/macho.rs]: Ok, then we should handle the None case manually.

```rust
let macho_id = match macho_file.mach_uuid() {
  Some(id) => format!("{:X}0", id.simple());,
  None => return Err(GetSymbolsError::InvalidInputError("Could not get mach uuid"))
};
```
or something similar.

File: src/elf.rs (hunk 14 above)
Comment (tinoism) [reply src/elf.rs]: `.map_err` needs a function as a first parameter. Do you mean by `let elf_file = buffer.map_err(ElfFile::parse, Err(GetSymbolsError::InvalidInputError(
        "Could not parse ELF header",
    )))?;`?

Comment (mstange) [reply src/elf.rs]: The function that you pass would be a lambda function that creates the error object. For example:

```rust
let elf_file = ElfFile::parse(buffer).map_err(|_| GetSymbolsError::InvalidInputError("Could not parse ELF header"))?;
```

or even

```rust
let elf_file = ElfFile::parse(buffer).map_err(|e| GetSymbolsError::ElfFileParseError(e))?;
```

File: src/error.rs (hunk 12 above)
Comment (mstange) [reply src/error.rs]: You can still get the error message out of the stored PDBError object later, can't you? Instead of calling `pdb_error.to_string()` in `GetSymbolsError::from<PDBError>`, you can call it later in `GetSymbolsError::fmt`.
In practice it won't really make a difference if this code is only ever used through the one WebAssembly entry function. But we may want to use the wasm-independent part of the code for other things in the future.

File: src/lib.rs [hunk 16]
```diff
@@ -126,14 +129,17 @@ pub fn get_compact_symbol_table(
     debug_data: &WasmMemBuffer,
     breakpad_id: &str,
     dest: &mut CompactSymbolTable,
-) -> bool {
+) -> JsValue {
```
Comment (mstange) [inline src/lib.rs]: I have reconsidered my position on the `dest` outparameter. It turns out my worries were unwarranted: wasm_bindgen has great support for returning types of the form `Result<SomeRustStruct, JsValue>`, so we should make use of that support.
Specifically: Can you please remove the `dest` parameter, and instead make this function return a value of type `Result<CompactSymbolTable, JsValue>`? Then we won't even need to use `JsValue::NULL` for the error case - when there is no error, we'll just return `Ok(table)`. You can probably even use `.map_err` here to turn the the `GetSymbolsError` into a `JsValue` (via a `GetSymbolsErrorJson`).

File: src/macho.rs [hunk 17]
```diff
@@ -1,10 +1,23 @@
...
-
-pub fn get_compact_symbol_table(buffer: &[u8], breakpad_id: &str) -> Option<CompactSymbolTable> {
-    let macho_file = MachOFile::parse(buffer).ok()?;
-    if format!("{:X}0", macho_file.mach_uuid()?.simple()) != breakpad_id {
-        return None;
+pub fn get_compact_symbol_table(buffer: &[u8], breakpad_id: &str) -> Result<CompactSymbolTable> {
+    let macho_file = MachOFile::parse(buffer).or(Err(GetSymbolsError::InvalidInputError(
+        "Could not parse Mach-O header",
+    )))?;
+    let macho_id = match macho_file.mach_uuid() {
```
Comment (mstange) [inline src/macho.rs]: Sorry for asking you to change this piece again, but there might actually be a better way to do this:

```rust
let macho_id = macho_file.mach_uuid().ok_or_else(|| GetSymbolsError::InvalidInputError("..."))?;
```

File: src/macho.rs [hunk 18]
```diff
@@ -1,10 +1,23 @@
+use crate::error::{GetSymbolsError, Result};
 use compact_symbol_table::CompactSymbolTable;
 use object::{MachOFile, Object};
-
-pub fn get_compact_symbol_table(buffer: &[u8], breakpad_id: &str) -> Option<CompactSymbolTable> {
-    let macho_file = MachOFile::parse(buffer).ok()?;
-    if format!("{:X}0", macho_file.mach_uuid()?.simple()) != breakpad_id {
-        return None;
+pub fn get_compact_symbol_table(buffer: &[u8], breakpad_id: &str) -> Result<CompactSymbolTable> {
+    let macho_file = MachOFile::parse(buffer).or(Err(GetSymbolsError::InvalidInputError(
```
Comment (mstange) [inline src/macho.rs]:
```rust
let macho_file = MachOFile::parse(buffer).or_else(|e| GetSymbolsError::MachOHeaderParseError(e)))?;
```

File: src/pdb.rs [hunk 19]
```diff
@@ -30,29 +34,37 @@ pub fn get_compact_symbol_table(pdb_data: &[u8], breakpad_id: &str) -> Option<Co
...
+            let info = pdb.module_info(&module)?;
+            let mut symbols = info.symbols()?;
+            while let Some(symbol) = symbols.next()? {
                 if let Ok(SymbolData::Procedure(ProcedureSymbol { offset, .. })) = symbol.parse() {
-                    let name = symbol.name().ok()?;
-                    hashmap
-                        .entry(offset.to_rva(&addr_map)?.0)
-                        .or_insert_with(|| Cow::from(name.to_string().into_owned()));
+                    let name = symbol.name()?;
+                    match offset.to_rva(&addr_map) {
```
Comment (mstange) [inline src/pdb.rs]: Let's use `.ok_or_else` as suggested in the other place here, too. And let's use the error string "Failed to map offset to rva using PDB file.".

File: src/error.rs [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+    error_msg: String,
+}
+
+impl GetSymbolsErrorJson {
+    pub fn from_error(err: GetSymbolsError) -> Self {
+        let error_type = err.enum_as_string().to_string();
+        let error_msg = err.to_string();
+        GetSymbolsErrorJson {
+            error_type,
+            error_msg,
```
Comment (mstange) [inline src/error.rs]: Don't need the variables. Let's write this as

```rust
GetSymbolsErrorJson {
    error_type: err.enum_as_string().to_string(),
    error_msg: err.to_string(),
}
```

File: src/lib.rs [hunk 21]
```diff
@@ -92,48 +95,58 @@ fn get_compact_symbol_table_impl(
...
-            dest.buffer = table.buffer;
-            true
+        Result::Ok(table) => {
+            let addr = table.addr;
+            let index = table.index;
+            let buffer = table.buffer;
+            Ok(CompactSymbolTable {
+                addr,
+                index,
+                buffer,
```
Comment (mstange) [inline src/lib.rs]: Let's avoid the local variables here the same way.

File: src/lib.rs [hunk 22]
```diff
@@ -92,48 +95,58 @@ fn get_compact_symbol_table_impl(
...
+                if table.is_err() {
+                    first_error = table.err();
+                } else {
+                    return table;
                 }
             }
-            Hint::PE => {
-                return pdb::get_compact_symbol_table(debug_data, breakpad_id);
+            if first_error.is_some() {
+                Err(first_error.unwrap())
```
Comment (mstange) [inline src/lib.rs]: Please find a way to avoid this unwrap. You could use `unwrap_or_else` and replace the entire if condition with:

```rust
Err(first_error.unwrap_or_else(|| GetSymbolsError::InvalidInputError(...)))
```

File: src/lib.rs [hunk 23]
```diff
@@ -92,48 +95,58 @@ fn get_compact_symbol_table_impl(
...
-                return macho::get_compact_symbol_table(binary_data, breakpad_id);
-            }
-            Hint::MachFat(_) => {
-                let multi_arch = mach::MultiArch::new(binary_data).ok()?;
-                for fat_arch in multi_arch.iter_arches().filter_map(Result::ok) {
-                    let arch_slice = fat_arch.slice(binary_data);
-                    if let Some(table) = macho::get_compact_symbol_table(arch_slice, breakpad_id) {
-                        return Some(table);
-                    }
+    let mut first_error = None;
```
Comment (mstange) [inline src/lib.rs]: Let's move this variable down into the scope where it's used.

File: src/lib.rs [hunk 24]
```diff
@@ -92,48 +95,58 @@ fn get_compact_symbol_table_impl(
...
+    match goblin::peek(&mut reader)? {
+        Hint::Elf(_) => elf::get_compact_symbol_table(binary_data, breakpad_id),
+        Hint::Mach(_) => macho::get_compact_symbol_table(binary_data, breakpad_id),
+        Hint::MachFat(_) => {
+            let multi_arch = mach::MultiArch::new(binary_data)?;
+            for fat_arch in multi_arch.iter_arches().filter_map(std::result::Result::ok) {
+                let arch_slice = fat_arch.slice(binary_data);
+                let table = macho::get_compact_symbol_table(arch_slice, breakpad_id);
+                if table.is_err() {
```
Comment (mstange) [inline src/lib.rs]: Another opportunity to use `match`:

```rust
match macho::get_compact_symbol_table(arch_slice, breakpad_id) {
    Ok(table) => return table,
    Err(e) => first_error = e,
}
```

Also, you're overwriting `first_error` even when it's already set, so a better name for the variable would be `last_error`. :)

**Review by mstange — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 25]
```diff
@@ -8,3 +8,6 @@ Cargo.lock
 
 # These are backup files generated by rustfmt
 **/*.rs.bk
+
+# Remove all the *.wasm files
+*.wasm
```
Comment (mstange) [inline .gitignore]: Oh, can you revert this change? We want to keep checking in profiler_get_symbols_bg.wasm.

File: Cargo.toml [hunk 26]
```diff
@@ -8,8 +8,11 @@ license = "MIT/Apache-2.0"
 pdb = "0.3.0"
 scroll = "0.9.2"
 uuid = "0.7.2"
-wasm-bindgen = "0.2.47"
 js-sys = "0.3.24"
+wasm-bindgen = {version = "0.2.47", features = ["serde-serialize"] }
```
Comment (mstange) [inline Cargo.toml]: Let's bump this to 0.2.48 because that's the version you've been building with. (That version difference is the main reason why there are so many lines changed in the `profiler_get_symbols.js` file, I think.)

File: Cargo.toml [hunk 27]
```diff
@@ -28,3 +31,8 @@ crate-type = ["cdylib"]
 
 [profile.release]
 lto = true
+
+#[dependencies.failure]
```
Comment (mstange) [inline Cargo.toml]: Please remove this commented out section.

File: src/elf.rs (hunk 15 above)
Comment (mstange) [reply src/elf.rs]: Oh, here I meant to say "use `.ok_or_else` in the places where you're using `.ok_or`.

File: src/pdb.rs [hunk 28]
```diff
@@ -30,29 +34,34 @@ pub fn get_compact_symbol_table(pdb_data: &[u8], breakpad_id: &str) -> Option<Co
...
+            let info = pdb.module_info(&module)?;
+            let mut symbols = info.symbols()?;
+            while let Some(symbol) = symbols.next()? {
                 if let Ok(SymbolData::Procedure(ProcedureSymbol { offset, .. })) = symbol.parse() {
-                    let name = symbol.name().ok()?;
+                    let name = symbol.name()?;
+                    let query = offset
+                        .to_rva(&addr_map)
+                        .ok_or_else(|| {
+                            GetSymbolsError::InvalidInputError("Cannot parse pdb hashmap")
```
Comment (mstange) [inline src/pdb.rs]: You did not apply my suggestion of a different error string.

**Review by mstange — CHANGES_REQUESTED** [review]
It looks like the file error.rs is missing?

In at least one of my previous comments I gave example code which used a non-existing error variant. This was meant as a request to add such a variant.

**Review by mstange — APPROVED** [review]
Looks great, thanks! I'll land it now and then make some small tweaks.
